core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -703,10 +708,12 @@ RexNode simplifyAnd2(List<RexNode> terms, List<RexNode> notTerms) {
     // Example #1. x AND y AND z AND NOT (x AND y)  - not satisfiable
     // Example #2. x AND y AND NOT (x AND y)        - not satisfiable
     // Example #3. x AND y AND NOT (x AND y AND z)  - may be satisfiable
-    for (RexNode notDisjunction : notTerms) {
-      final List<RexNode> terms2 = RelOptUtil.conjunctions(notDisjunction);
-      if (terms.containsAll(terms2)) {
-        return rexBuilder.makeLiteral(false);
+    if (unknownAsFalse) {
```

Comment:
This method should only be called when ```unknownAsFalse``` is false, right? It is very confusing that this method checks the value of this boolean, given that we should only be in it if ```unknownAsFalse``` is false.
It would be more clear to change the boolean value before calling ```simplifyAnd```, that would fix the issue.

Reply:
Just to make sure my comment is understood: since we have two variants for the method, ideal solution is to prevent any method from calling ```simplifyAnd2``` if ```unknownAsFalse``` is true.